### PR TITLE
cephadm: document extended gdb debugging techniques

### PR DESCRIPTION
**Pull request by phlogistonjohn**
Document some techniques for debugging with cephadm. Add a simple flag to cephadm shell to help enable this.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [x] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), <s>opened tracker ticket</s> incredibly trivial code change
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by phlogistonjohn** [comment]
FYI @mattbenjamin

File: doc/cephadm/troubleshooting.rst [hunk 1]
```diff
@@ -390,3 +390,78 @@ You can then run debugger commands at gdb's prompt.
...
+When using ``cephadm shell``, like in the example above, the changes made to
+the container the shell command spawned are ephemeral. Once the shell session
+exits all of the files that were downloaded and installed are no longer
+available. One can simply re-run the same commands every time ``cephadm shell``
+is invoked, but in order to save time and resources one can create a new
+container image and use it for repeated debugging sessions.
+
+In the following example we create a simple file for constructing the
+container image. The command below uses podman but it should work correctly
+is ``podman`` is replaced with ``docker``.
```
Comment (adk3798) [inline doc/cephadm/troubleshooting.rst:407]: `is podman` -> `if podman`

File: doc/cephadm/troubleshooting.rst [hunk 2]
```diff
@@ -390,3 +390,78 @@ You can then run debugger commands at gdb's prompt.
...
+
+Now, we can run our debugging container with ``sh /tmp/debug.sh``. Within the shell
+we can run commands such as ``ps`` to get the PID of the MGR process. In the following
+example this will be ``2``. Running gdb, we can now attach to the running process:
+
+.. prompt:: bash (gdb)
+
+   attach 2
+   info threads
+   bt
```
Comment (adk3798) [inline doc/cephadm/troubleshooting.rst:467]: Does this all work in cases where the process is quickly crashing, or is it that you're meant to use the coredump strategy for those cases and this is only for processes that will actually keep running.

**Review by adk3798 — APPROVED** [review]
Content LGTM

File: doc/cephadm/troubleshooting.rst (hunk 1 above)
Comment (phlogistonjohn) [reply doc/cephadm/troubleshooting.rst]: Fixed.

File: doc/cephadm/troubleshooting.rst (hunk 2 above)
Comment (phlogistonjohn) [reply doc/cephadm/troubleshooting.rst]: I am not entirely sure what you mean. If you're asking if you should try to attach to a process this way if a process crashes shortly after it is started, then the answer is -- it depends. :-)
Partly it depends on how quickly it crashes after start. You can also try techniques like attaching to the parent process before the process you want to debug forks and following the fork, but in the general case, I would suggest that debugging using the core would be simpler.

File: doc/cephadm/troubleshooting.rst [hunk 3]
```diff
@@ -335,33 +335,133 @@ Deploy the daemon::
...
+
+
+To enable coredump handling, run:
 
 .. prompt:: bash #
 
   ulimit -c unlimited
 
-Core dumps will now be written to ``/var/lib/systemd/coredump``.
+Core dumps will be written to ``/var/lib/systemd/coredump``.
```
Comment (zdover23) [inline doc/cephadm/troubleshooting.rst:352]: I think the present tense is just a little bit easier on the reader.
s/will be/are/

File: doc/cephadm/troubleshooting.rst [hunk 4]
```diff
@@ -335,33 +335,133 @@ Deploy the daemon::
...
+This will persist until the system is rebooted.
 
 .. note::
 
   Core dumps are not namespaced by the kernel, which means
   they will be written to ``/var/lib/systemd/coredump`` on
   the container host. 
 
-Now, wait for the crash to happen again. To simulate the crash of a daemon, run e.g. ``killall -3 ceph-mon``.
+Now, wait for the crash to happen again. To simulate the crash of a daemon, run
```
Comment (zdover23) [inline doc/cephadm/troubleshooting.rst:361]: s/Now, wait/Wait/

File: doc/cephadm/troubleshooting.rst [hunk 5]
```diff
@@ -335,33 +335,133 @@ Deploy the daemon::
...
+e.g. ``killall -3 ceph-mon``.
+
+
+Running the Debugger with cephadm
+----------------------------------
+
+Running a single debugging session
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+One can initiate a debugging session using the ``cephadm shell`` command.
```
Comment (zdover23) [inline doc/cephadm/troubleshooting.rst:371]: s/One can initiate/Initiate/

File: doc/cephadm/troubleshooting.rst [hunk 6]
```diff
@@ -335,33 +335,133 @@ Deploy the daemon::
...
+    #0  0x00007fa9117383fc in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
+    #1  0x00007fa910d7f8f0 in std::condition_variable::wait(std::unique_lock<std::mutex>&) () from /lib64/libstdc++.so.6
+    #2  0x00007fa913d3f48f in AsyncMessenger::wait() () from /usr/lib64/ceph/libceph-common.so.2
+    #3  0x0000563085ca3d7e in main ()
+
+
+Running repeated debugging sessions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When using ``cephadm shell``, like in the example above, the changes made to
```
Comment (zdover23) [inline doc/cephadm/troubleshooting.rst:398]: s/like in the example below/as in the example below/

File: doc/cephadm/troubleshooting.rst [hunk 7]
```diff
@@ -335,33 +335,133 @@ Deploy the daemon::
...
+the container the shell command spawned are ephemeral. Once the shell session
```
Comment (zdover23) [inline doc/cephadm/troubleshooting.rst:399]: s/the container the shell command spawned/the container that the shell command spawned/

Comment (zdover23) [inline doc/cephadm/troubleshooting.rst:399]: s/Once the shell/After the shell/

File: doc/cephadm/troubleshooting.rst [hunk 8]
```diff
@@ -335,33 +335,133 @@ Deploy the daemon::
...
+exits all of the files that were downloaded and installed are no longer
```
Comment (zdover23) [inline doc/cephadm/troubleshooting.rst:400]: s/exits all/exits, all/

Comment (zdover23) [reply doc/cephadm/troubleshooting.rst:400]: Do not concern yourself with making these usage changes. I will make them in a future PR, after this material is merged.

Comment (phlogistonjohn) [reply doc/cephadm/troubleshooting.rst:400]: OK, whatever you prefer. I think all the suggestions are reasonable and happy to apply them. I'm also lazy and happy to let other people clean up after I make a mess. ;-)

**Comment by phlogistonjohn** [comment]
jenkins test windows

**Review by zdover23 — APPROVED** [review]

**Comment by zdover23** [comment]
https://github.com/ceph/ceph/pull/54220 - Reef backport
https://github.com/ceph/ceph/pull/54221 - Quincy backport

**Comment by zdover23** [comment]
See https://github.com/ceph/ceph/pull/54281.

**Pull request merged**